src/sagemaker_sklearn_extension/preprocessing/encoders.py

```diff
@@ -620,3 +621,248 @@ def inverse_transform(self, X):
+        except TypeError:
+            pass
+
+    def _feature_iterator(self, X):
+        """Create a lazy iterator over the columns of `X`. Each element
+        returned by the iterator is a numpy vector.
+        """
+        if hasattr(X, 'values'):
+            X_raw = X.values
+        else:
+            X_raw = X
+        dims = X_raw.shape
+        if len(dims) == 1:  # vector
+            feature_iterator = (X_raw,)
+        else:  # 2d array
+            p = dims[1]
+            # selection of columns: either all or only given ones
+            sel = self.feature_indices if self.feature_indices else range(p)
+            assert all(0 <= i < p for i in sel), \
+                WOEAsserts.FEAT_OOR
+            feature_iterator = (X_raw[:, i] for i in sel)
+
+        return feature_iterator
+
+    def _woe(self, x):
+        """Return the categories for a feature vector `x` as well as the
+        corresponding weight of evidence value for each of those categories.
+        """
+        self._check_no_missing(x)
+        cat_x = np.unique(x)
+
+        # Computation of the Weight of Evidence
+        #
+        #   woe_c = log( { #(y==0 | c) + α / #(y==1 | c)  + α } *
+        #                { #(y==1) + β / #(y==0) + β } )
+        #
+        # where β = 2α if laplace == True, 0 otherwise. The second factor
+        # can be computed once, call it `r10` then
+        #
+        #   woe_c = log( r10 * ratio(c) )
+        #
+        # where
+        #
+        #   ratio(c) = { #(y==0 | c) + α } / { #(y==1 | x==c) + α }
+        #
+
+        def ratio(c):
+            x_c = x == c
+            # retrieve the number of (y == 0 | x == c) and same for y == 1
+            y_0_c = sum(np.logical_and(self.mask_y_0_, x_c))
+            y_1_c = sum(np.logical_and(self.mask_y_1_, x_c))
+            # compute the ratio with regularization for 0 events
+            return (y_0_c + self.alpha) / (y_1_c + self.alpha)
+
+        # computation of woe possibly using Laplace smoothing (beta factor)
+        r10 = (self.count_y_[1] + self.beta_) / (self.count_y_[0] + self.beta_)
+        woe = np.log(r10 * np.array([ratio(c) for c in cat_x]))
+
+        # encoder from unique values of x to index
+        codex = {c: i for (i, c) in enumerate(cat_x)}
+
+        return (codex, woe)
+
+    def _encode(self, t):
+        i, x = t
+        codex, woe = self.woe_pairs_[i]
+        if self.binning:
+            x = self.binners_[i].transform(x.reshape(-1, 1)).flatten()
+
+        return [woe[codex[xi]] for xi in x]
+
+    def fit(self, X, y):
+        """Fit Weight of Evidence encoder to `X` and `y`.
+
+        Parameters
+        ----------
+        X: array-like, shape (n_samples, n_features)
+            The data to encode.
+
+        y: array-like, shape (n_samples,)
+            The binary target vector.
+
+        Returns
+        -------
+        self: WOEEncoder.
+        """
+        X, y = check_X_y(X, y)
```

Comment:
I recommend using `sklearn.utils.check_array`, especially if the X and y inputs are "array-like". There should also be an option for check_array to not accept missing values, eliminating the need for L817

Reply:
https://scikit-learn.org/stable/modules/generated/sklearn.utils.check_X_y.html?highlight=check_x_y#sklearn.utils.check_X_y

* effectively calls `check_array` (on both `X` and `y`) (and you're right that it checks for missing so I can scrap that)